website/docs/opengl/c1/2-hello-quad.md

```diff
@@ -0,0 +1,524 @@
+
+Congrats! You've done your first thing in OpenGL! Didn't work? Check the [source code](../sources/1.2.2-clear-window.html) for this section here.
+
+But - how does OpenGL know to color the window blue? Sure, we've told it we want blue - but we only did that once... How does it know to keep using blue?
+
+Well, this is because OpenGL is a **state machine**.
+
+## OpenGL: The state machine
+Before we can continue further, you'll have to get used to the concept that OpenGL is a **state machine**. It's really hard to understand OpenGL if you don't understand this, so read closely.
+
+### What is a state machine?
+At its core, a state machine holds... state. You set the state of something, and it retains that state until you change it.
+
+This is exactly how OpenGL works. Once you set something, it will remain set until you change it. This counts for everything in OpenGL. Clear color, binding objects, etc etc, everything goes through the state machine. You can *manipulate* the current state, however you have to be wary at all times of what part of the state you are changing. Change the wrong thing and suddenly you program might not work!
+
+So, this explains why clearing the window works. You set the clear color, and it remains *as* the clear color until you change it.
+
+Now that we've got that, let's move on!
+
+## Vertex Array Objects (VAOs)
+A unique feature that modern OpenGL has is what is known as a Vertex Array Object, or VAO. It stores the state of various parameters, such as which buffers you wish to use, and the shader inputs.
+
+This is a required feature of modern OpenGL. You **must** have a VAO bound, otherwise your application won't work.
+
+Let's create the VAO!
+
+At the top of your class, add the following:
+
+```cs
+private static uint _vao;
+```
+
+Then, in your `OnLoad` method, add:
+
+```cs
+_vao = _gl.GenVertexArray();
+_gl.BindVertexArray(_vao);
+```
+
+What's going on here? First, we generate the VAO. Then, before we can use, or update it, we need to **bind** it.
+
+Binding is a term you will hear a lot with OpenGL, and in these tutorials too. Binding essentially means updating the current OpenGL state with the given object. So, in this example, when we bind the VAO, we update OpenGL's state so that the VAO we are currently using is now the VAO we have just created. If you don't bind an object before updating/using it, you will use or update the previously bound object instead, which can cause problems and headaches later.
+
+We're now ready to give the VAO some data. But first, we need a **vertex buffer**.
+
+## Vertex Buffer Objects (VBOs)
+Here is the part where we actually give the GPU some data it can work with. For 95% of graphics applications, a vertex buffer is **required** to display anything on screen.
+
+### What is a vertex buffer?
+Before we can continue, it's handy to know what a vertex buffer is, and what it does.
+
+Let's define both vertex, and buffer.
+
+* Vertex - A point where two lines, or edges meet.
```

Comment:
(This is just a note, I'm not asking for this to be changed, I like it because it's simple to understand. I just want to comment this to make sure you are aware that the OpenGL concept of "vertex" is actually much more complex)

Note: A vertex in OpenGL isn't necessarily this. You can define the vertex attributes you want and use them however you wish to. For example, a vertex might have just a "texture coordinates" attribute, and then the vertex shader gets the position (to assign to `gl_Position`) from a texture. Another example is a small heightmap test project of mine, in which I had an empty VAO; No vertex attributes whatsoever! The vertex shader instead used `gl_VertexID` to organize the vertex into a flat XZ mesh pattern, and then the Y position came from reading the heightmap texture.

Reply:
Yeah, and this may be explored in later tutorials, I've been meaning to get into heightmaps and stuff myself too. For the purposes of this tutorial though, the vertex definition used here is helpful enough for the reader to understand.